### ISSUE TITLE
ENH: Extend `linprog(method='highs')` to use `x0` argument as initial guess to speed up optimization

### ISSUE THREAD

**Issue by carlosgmartin** [scipy/scipy#21591]
### Is your feature request related to a problem? Please describe.

The [`linprog`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html) documentation contains the following description of the `x0` argument:

> Guess values of the decision variables, which will be refined by the optimization algorithm. <ins>This argument is currently used only by the ‘revised simplex’ method</ins>, and can only be used if x0 represents a basic feasible solution.

Currently, there is no way to pass an initial guess of the solution (this could be, for example, the solution to a similar problem) to [`linprog(method='highs')`](https://docs.scipy.org/doc/scipy/reference/optimize.linprog-highs.html) to speed up optimization.

### Describe the solution you'd like.

Extend the implementation of [`linprog(method='highs')`](https://docs.scipy.org/doc/scipy/reference/optimize.linprog-highs.html) to use the `x0` argument to [`linprog`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html) as an initial guess.

**Comment by nickodell** [comment]
Could this be done via the setSolution method in HiGHS? https://ergo-code.github.io/HiGHS/dev/guide/further/#hot-start

If so, do we need to do any validation of the provided solution? e.g. does HiGHS assume that the provided solution does not violate constraints?

It mentions that support for this is solver-specific. Do the solvers we use support this?

**Comment by M-I-Dx** [comment]
Any update on this?

**Comment by nickodell** [comment]
> Any update on this?

No, if there were an update to this, it would have been posted to this issue. Are you interested in working on this? If you are, I could give you some pointers about where to start, depending on your experience level with Python / C++.

**Comment by M-I-Dx** [comment]
I am definitely interested in working on this. I am fairly proficient in Python but do not have much experience in Cpp.

**Comment by nickodell** [comment]
> I am fairly proficient in Python but do not have much experience in Cpp.

Okay - this project probably won't involve writing any C++, but it might involve reading some, in the research step.

>I am definitely interested in working on this.

I would suggest starting with two items.

1. Creating a build environment. https://docs.scipy.org/doc/scipy/building/index.html

   You will need this to test any changes to SciPy.
3. Research. I mentioned before that "It mentions that support for this is solver-specific. Do the solvers we use support this?"
   The two things that are required before we add something like this is

    1. Some documentation or code that shows that at least one of the solvers we use, ipm and simplex, support this.
    2. Some kind of proof of concept where providing a good x0 makes the linprog solve faster.

   For the first research item, I would suggest looking at the HiGHs documentation, reading the HiGHs source code around setSolution, or asking questions on the HiGHs issue tracker.

   For the second research item, I would suggest using [highspy](https://pypi.org/project/highspy/), which provides a low-level interface into HiGHs. We use version 1.8.0, so I would suggest using the same version to ensure that your experiment is valid. To figure out how to call HiGHs with this low-level interface, I would suggest that you read either SciPy's [highs driver](https://github.com/scipy/scipy/blob/main/scipy/optimize/_highspy/_highs_wrapper.py#L9), or the [HiGHs examples](https://ergo-code.github.io/HiGHS/dev/interfaces/python/example-py/). I would suggest that you try to come up with an example of a linear optimization problem where calling `setSolution` on the model object causes it to be solved faster. Given that you are asking about this on this issue tracker, I bet you already have an idea of a problem where this could benefit. : )

   You might run into an issue where presolve times dominate run time, and that makes it hard to measure the effect of setSolution. For that reason, I would suggest running this optimization with and without presolve, and measuring the effect of providing setSolution on each.

If you need me to elaborate on any of that, or have questions about SciPy or SciPy's development process, please ask!

**Comment by nickodell** [comment]
@jajhall I wanted to ask you if you thought that providing an initial guess for a solve is technically feasible / useful for reducing solve time within HiGHS. If it is, it would be good to expose this feature to our users.

Here is what I am specifically interested in:

> Could this be done via the setSolution method in HiGHS? https://ergo-code.github.io/HiGHS/dev/guide/further/#hot-start
> 
> If so, do we need to do any validation of the provided solution? e.g. does HiGHS assume that the provided solution does not violate constraints?
> 
> It mentions that support for this is solver-specific. Do the solvers we use support this?

**Comment by jajhall** [comment]
A solution is supplied as you describe, and it's checked for feasibility. If it's not feasible then it's not used.

It can be used by the MIP solver and simplex solver, but it's better to give the latter an advanced basis. The value for the MIP solver is variable , and unlikely to be great. Even if it's optimal, the dual bound still has to be established before optimality can be declared.

**Comment by jajhall** [comment]
A solution is supplied as you describe, and it's checked for feasibility. If it's not feasible then it's not used.

It can be used by the MIP solver and simplex solver, but it's better to give the latter an advanced basis. The value for the MIP solver is variable , and unlikely to be great. Even if it's optimal, the dual bound still has to be established before optimality can be declared.
